File: fsm/realtime.py
# Copyright (c) 2026, KodLyft and contributors
# For license information, please see license.txt

import frappe

SERVICE_JOB_EVENT = "fsm_service_job_update"
TECHNICIAN_EVENT = "fsm_technician_update"


def on_service_job_update(doc, method=None):
	"""Broadcast a compact Service Job summary to dispatch dashboards."""
	frappe.publish_realtime(
		SERVICE_JOB_EVENT,
		{
			"name": doc.name,
			"customer_name": doc.customer_name,
			"status": doc.status,
			"priority": doc.priority,
			"scheduled_date": str(doc.scheduled_date) if doc.scheduled_date else None,
			"primary_technician": doc.primary_technician,
			"sla_breached": doc.sla_breached,
		},
		after_commit=True,
	)


def on_technician_update(doc, method=None):
	"""Broadcast technician availability / location changes to the dispatch board."""
	frappe.publish_realtime(
		TECHNICIAN_EVENT,
		{
			"name": doc.name,
			"technician_name": doc.technician_name,
			"status": doc.status,
			"clocked_in": doc.clocked_in,
			"last_seen_latitude": doc.last_seen_latitude,
			"last_seen_longitude": doc.last_seen_longitude,
			"last_seen_on": str(doc.last_seen_on) if doc.last_seen_on else None,
		},
		after_commit=True,
	)

File: frontend/console/src/lib/realtime.ts
import { Realtime } from "@kodlyft/api";

export const SERVICE_JOB_EVENT = "fsm_service_job_update";
export const TECHNICIAN_EVENT = "fsm_technician_update";
export const JOB_MESSAGE_EVENT = "fsm_job_message";

export interface JobMessageEvent {
	service_job: string;
	author_role: string;
	author_name: string | null;
	message: string;
	creation: string;
}

export interface ServiceJobEvent {
	name: string;
	customer_name: string;
	status: string;
	priority: string;
	scheduled_date: string | null;
	primary_technician: string | null;
	sla_breached: number;
}

export interface TechnicianEvent {
	name: string;
	technician_name: string;
	status: string;
	clocked_in: number;
	last_seen_latitude: number | null;
	last_seen_longitude: number | null;
	last_seen_on: string | null;
}

let instance: Realtime | null = null;

export function realtime(): Realtime {
	if (!instance) instance = new Realtime({ url: "" });
	return instance;
}
